### PR TITLE
feature/unread_conversations

### PR DESCRIPTION
**Pull request by Rodrigo-Miranda-0**
#### Board:
* [Ticket #WDH1tFxm](https://trello.com/c/WDH1tFxm/13-3-as-a-user-i-should-be-able-to-see-the-unread-conversations-count-so-i-can-easily-identify-them)
---
#### Description:

* Allows the user to view unread message amount in the corresponding conversations
---
#### Notes:
*
---
#### Tasks:
- [x] View how many unread messages a user has
---
#### Risk:
*
---
#### Preview:

### CODE REVIEW

File: app/controllers/api/v1/messages_controller.rb [hunk 1]
```diff
@@ -9,14 +9,17 @@ def create
       end
 
       def index
-        conversation = Conversation.find(params[:conversation_id])
-        raise InvalidConversationError unless conversation.user1_id == @current_user.id || conversation.user2_id == @current_user.id
-
-        render json: { messages: conversation.messages.page(params[:page]) }, status: :ok
+        @messages = IndexMessageService.new.index(params[:conversation_id], current_user, params[:page])
```
Comment (jpascual1994) [inline app/controllers/api/v1/messages_controller.rb]: I would name the service MessageIndexService so it starts with the model instead of the action

File: app/controllers/api/v1/messages_controller.rb [hunk 2]
```diff
@@ -9,14 +9,17 @@ def create
...
       rescue InvalidConversationError => e
         render json: { error: e.message }, status: :unprocessable_entity
       end
 
+      def unread_messages
+        conversations = Conversation.where(user1: current_user).or(Conversation.where(user2: current_user))
+        unread_messages = conversations.map { |conversation| { id: conversation.id, count: conversation.messages.where(user: conversation.opposite_user(current_user)).where(read: false).count } }
```
Comment (jpascual1994) [inline app/controllers/api/v1/messages_controller.rb]: maybe you can move `conversation.messages.where(user: conversation.opposite_user(current_user)).where(read: false).count` to a method on the conversation model and use it.
Also it's worth using jbuilder for this view

File: app/services/api/v1/index_message_service.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+        messages = conversation.messages.page(page)
+        read_messages(messages)
+        messages
+      end
+
+      private
+
+      def read_messages(messages)
+        messages.each { |message| message.update(read: true) }
```
Comment (jpascual1994) [inline app/services/api/v1/index_message_service.rb]: I think this should update the messages from the opposite user so you don't mark as read the message sent by yourself

File: app/controllers/api/v1/messages_controller.rb [hunk 4]
```diff
@@ -20,7 +20,7 @@ def index
       private
 
       def message_params
-        params.require(:message).permit(:content)
+        params.permit(:content)
```
Comment (jpascual1994) [inline app/controllers/api/v1/messages_controller.rb]: why did you remove the require?

File: config/routes.rb [hunk 5]
```diff
@@ -9,7 +9,11 @@
       resources :targets, only: %i[index create destroy]
       resources :user, only: :update
       resources :conversations, only: :index
-      resources :messages, only: %i[index create]
+      resources :messages, only: %i[index create] do
+        collection do
+          get :unread_messages
+        end
```
Comment (jpascual1994) [inline config/routes.rb]: Are we using this endpoint?

File: db/migrate/20221124210109_create_messages.rb [hunk 6]
```diff
@@ -2,11 +2,11 @@ class CreateMessages < ActiveRecord::Migration[7.0]
...
     create_table :messages do |t|
       t.string :content
-      t.references :user, null: false, foreign_key: true
-      t.references :conversation, null: false, foreign_key: true
+      t.boolean :read, default: false
+      t.references :user, null: false
+      t.references :conversation, null: false
       t.timestamps
 
-      t.index [:user_id, :conversation_id], unique: true
```
Comment (jpascual1994) [inline db/migrate/20221124210109_create_messages.rb]: These changes should be on a new migration, remember we shouldn't edit migrations

File: db/migrate/20221205142435_edit_message_table.rb [hunk 7]
```diff
@@ -1,5 +0,0 @@
-class EditMessageTable < ActiveRecord::Migration[7.0]
-  def change
-    remove_index :messages, [:user_id, :conversation_id]
```
Comment (jpascual1994) [inline db/migrate/20221205142435_edit_message_table.rb]: Why are we removing a migration here?

File: spec/requests/api/v1/messages/index_spec.rb [hunk 8]
```diff
@@ -24,7 +24,6 @@
 
     context 'rest of the pages' do
       let(:page) { 2 }
-
```
Comment (jpascual1994) [inline spec/requests/api/v1/messages/index_spec.rb]: Usually we let an empty space between the last let and the first context/it

**Review by jpascual1994 — APPROVED** [review]

File: app/models/conversation.rb [hunk 9]
```diff
@@ -28,4 +28,16 @@ class Conversation < ApplicationRecord
   def recipient_user(current_user)
     current_user == user1 ? user2 : user1
   end
+
+  def opposite_user(user)
+    user1 == user ? user2 : user1
+  end
+
+  def unread_messages(current_user)
```
Comment (juanmferreira93) [inline app/models/conversation.rb:36]: it's not `current_user` a helper? So I assume you could not send it as params. Device provide it for you

File: app/services/api/v1/message_index_service.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
+module Api
+  module V1
+    class MessageIndexService
+      def index(conversation_id, current_user, page)
```
Comment (juanmferreira93) [inline app/services/api/v1/message_index_service.rb:4]: same here, please try without send it.. it should work

**Review by juanmferreira93 — APPROVED** [review]
